File: cloud-run/modules/notebooklm_sync.py
"""
NotebookLM同期モジュール
処理済みファイルのOCRテキストを年度別累積ドキュメントに追記する
Drive APIのみを使用してGoogleドキュメントを作成・更新
"""
import logging
from typing import Optional
from datetime import datetime
from io import BytesIO
from googleapiclient.http import MediaIoBaseUpload
from modules.drive_client import DriveClient
from config.settings import (
    FOLDER_IDS,
    NOTEBOOKLM_SYNC_CATEGORIES
)

logger = logging.getLogger(__name__)


class NotebookLMSync:
    """NotebookLM用シャドウドキュメント同期クラス"""
    
    # 同期済みマーカー（Driveプロパティに設定）
    PROCESSED_MARKER = 'notebooklm_synced'
    
    def __init__(self, drive_client: DriveClient):
        """
        初期化
        
        Args:
            drive_client: DriveClientインスタンス
        """
        self.drive_client = drive_client
    
    def should_sync(self, category: str) -> bool:
        """
        このカテゴリが同期対象かどうかをチェック
        
        Args:
            category: カテゴリ名
            
        Returns:
            同期対象の場合True
        """
        return category in NOTEBOOKLM_SYNC_CATEGORIES
    
    def sync_file(
        self,
        file_id: str,
        file_name: str,
        category: str,
        ocr_text: str,
        date_str: str,
        fiscal_year: int
    ) -> bool:
        """
        ファイルのOCRテキストを累積ドキュメントに追記
        
        Args:
            file_id: 元ファイルのID
            file_name: 元ファイル名
            category: カテゴリ名
            ocr_text: OCRで抽出されたテキスト
            date_str: YYYYMMDD形式の日付
            fiscal_year: 年度
            
        Returns:
            成功したかどうか
        """
        if not self.should_sync(category):
            logger.info(f"カテゴリ {category} は同期対象外です")
            return False
        
        try:
            # 日付をフォーマット
            formatted_date = self._format_date(date_str)
            
            # 累積ドキュメントを取得または作成
            doc_id = self._get_or_create_accumulated_doc(fiscal_year)
            if not doc_id:
                logger.error(f"累積ドキュメント取得/作成失敗: {fiscal_year}年度")
                return False
            
            # ドキュメントに追記（カテゴリ名を含める）
            entry_text = self._format_entry(formatted_date, file_name, file_id, ocr_text, category)
            if not self._append_to_doc(doc_id, entry_text):
                logger.error(f"ドキュメント追記失敗: {doc_id}")
                return False
            
            # 元ファイルに同期済みマーカーを設定
            self._mark_as_synced(file_id)
            
            logger.info(f"NotebookLM同期完了: {file_name} → {fiscal_year}年度_全記録")
            return True
            
        except Exception as e:
            logger.error(f"NotebookLM同期エラー: {e}")
            return False
    
    def _format_date(self, date_str: str) -> str:
        """YYYYMMDD形式をYYYY/MM/DD形式に変換"""
        if not date_str or len(date_str) != 8:
            return datetime.now().strftime('%Y/%m/%d')
        return f"{date_str[:4]}/{date_str[4:6]}/{date_str[6:8]}"
    
    def _format_entry(
        self,
        formatted_date: str,
        file_name: str,
        file_id: str,
        ocr_text: str,
        category: str
    ) -> str:
        """エントリテキストをフォーマット（カテゴリ名を含む）"""
        file_url = f"https://drive.google.com/file/d/{file_id}/view"
        
        entry = f"""

========================================
📄 {formatted_date} - [{category}] {file_name}
🔗 {file_url}
========================================

{ocr_text}

"""
        return entry
    
    def _get_or_create_accumulated_doc(self, fiscal_year: int) -> Optional[str]:
        """
        年度別統合ドキュメントを取得または作成
        全カテゴリを1つのドキュメントに統合（NotebookLM用）
        
        Args:
            fiscal_year: 年度
            
        Returns:
            ドキュメントID
        """
        sync_folder_id = FOLDER_IDS.get('NOTEBOOKLM_SYNC')
        if not sync_folder_id:
            logger.error("NOTEBOOKLM_SYNCフォルダIDが設定されていません")
            return None
        
        # ドキュメント名（年度のみ、全カテゴリ統合）
        doc_name = f"{fiscal_year}年度_全記録"
        
        # 既存のドキュメントを検索（同期フォルダ直下）
        doc_id = self._find_doc_by_name(doc_name, sync_folder_id)
        if doc_id:
            return doc_id
        
        # 新規作成（同期フォルダ直下）- Drive APIを使用
        return self._create_unified_doc(doc_name, sync_folder_id, fiscal_year)
    
    def _find_doc_by_name(self, doc_name: str, parent_id: str) -> Optional[str]:
        """
        フォルダ内でGoogleドキュメントを名前で検索
        
        Args:
            doc_name: ドキュメント名
            parent_id: 親フォルダID
            
        Returns:
            ドキュメントID（見つからない場合はNone）
        """
        try:
            query = (
                f"name='{doc_name}' and '{parent_id}' in parents "
                f"and mimeType='application/vnd.google-apps.document' and trashed=false"
            )
            results = self.drive_client.service.files().list(
                q=query,
                fields='files(id, name)'
            ).execute()
            
            files = results.get('files', [])
            if files:
                return files[0]['id']
            return None
        except Exception as e:
            logger.error(f"ドキュメント検索エラー: {e}")
            return None
    
    def _create_unified_doc(
        self,
        doc_name: str,
        parent_id: str,
        fiscal_year: int
    ) -> Optional[str]:
        """
        Drive APIを使用して新しい統合ドキュメントを作成（全カテゴリ用）
        
        Args:
            doc_name: ドキュメント名
            parent_id: 親フォルダID
            fiscal_year: 年度
            
        Returns:
            作成されたドキュメントID
        """
        try:
            # ヘッダーテキスト
            header_text = f"""# {fiscal_year}年度 全記録

このドキュメントは NotebookLM 用に自動生成された書類OCRテキストの統合ファイルです。
各エントリには [カテゴリ名] が付与されています。

---

"""
            # Drive APIでGoogleドキュメントを作成
            file_metadata = {
                'name': doc_name,
                'mimeType': 'application/vnd.google-apps.document',
                'parents': [parent_id]
            }
            
            # 空のドキュメントを作成
            doc = self.drive_client.service.files().create(
                body=file_metadata,
                fields='id'
            ).execute()
            
            doc_id = doc.get('id')
            
            # ヘッダーを追加
            self._append_to_doc(doc_id, header_text)
            
            # オーナー権限をユーザーに転送（サービスアカウントの容量制限回避）
            # NOTE: transferOwnership=True は role='owner' の場合に必須
            from config.settings import NOTEBOOKLM_OWNER_EMAIL
            if NOTEBOOKLM_OWNER_EMAIL:
                try:
                    self.drive_client.service.permissions().create(
                        fileId=doc_id,
                        body={
                            'role': 'owner',
                            'type': 'user',
                            'emailAddress': NOTEBOOKLM_OWNER_EMAIL
                        },
                        transferOwnership=True
                    ).execute()
                    logger.info(f"ドキュメントのオーナー権限を転送しました: {NOTEBOOKLM_OWNER_EMAIL}")
                except Exception as e:
                    logger.warning(f"オーナー権限の転送に失敗しました（容量制限に注意）: {e}")

            logger.info(f"統合ドキュメント作成: {doc_name}")
            return doc_id
            
        except Exception as e:
            logger.error(f"ドキュメント作成エラー: {e}")
            return None
    
    def _append_to_doc(self, doc_id: str, text: str) -> bool:
        """
        ドキュメントの末尾にテキストを追記
        Drive APIのexport/updateを使用
        
        Args:
            doc_id: ドキュメントID
            text: 追記するテキスト
            
        Returns:
            成功したかどうか
        """
        try:
            # 現在のドキュメント内容をテキストとしてエクスポート
            current_content = self.drive_client.service.files().export(
                fileId=doc_id,
                mimeType='text/plain'
            ).execute()
            
            # 既存の内容がbytesの場合はデコード
            if isinstance(current_content, bytes):
                current_content = current_content.decode('utf-8')
            
            # 新しい内容を追加
            new_content = current_content + text
            
            # テキストファイルとしてアップロード（Googleドキュメントに変換）
            media = MediaIoBaseUpload(
                BytesIO(new_content.encode('utf-8')),
                mimetype='text/plain',
                resumable=True
            )
            
            self.drive_client.service.files().update(
                fileId=doc_id,
                media_body=media
            ).execute()
            
            return True
            
        except Exception as e:
            logger.error(f"ドキュメント追記エラー: {e}")
            return False
    
    def _mark_as_synced(self, file_id: str):
        """
        ファイルを同期済みとしてマーク
        
        Args:
            file_id: ファイルID
        """
        try:
            # Drive APIのpropertiesを使用
            self.drive_client.service.files().update(
                fileId=file_id,
                body={
                    'properties': {
                        self.PROCESSED_MARKER: 'true'
                    }
                }
            ).execute()
        except Exception as e:
            logger.warning(f"同期マーキングエラー: {e}")
    
    def is_already_synced(self, file_id: str) -> bool:
        """
        ファイルが既に同期済みかチェック
        
        Args:
            file_id: ファイルID
            
        Returns:
            同期済みの場合True
        """
        try:
            file = self.drive_client.service.files().get(
                fileId=file_id,
                fields='properties'
            ).execute()
            
            properties = file.get('properties', {})
            return properties.get(self.PROCESSED_MARKER) == 'true'
            
        except Exception as e:
            logger.warning(f"同期状態チェックエラー: {e}")
            return False

File: cloud-run-go/internal/service/notebooklm_sync.go
package service

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"sync"
	"time"

	"github.com/leo-sagawa/homedocmanager/internal/config"
	"google.golang.org/api/docs/v1"
	"google.golang.org/api/drive/v3"
	"google.golang.org/api/googleapi"
)

// NotebookLMSync はNotebookLM同期サービス
type NotebookLMSync struct {
	driveClient *DriveClient
	mu          sync.Mutex
}

const processedMarker = "notebooklm_synced"

// NewNotebookLMSync は新しいNotebookLMSyncを作成
func NewNotebookLMSync(ctx context.Context, driveClient *DriveClient) (*NotebookLMSync, error) {
	return &NotebookLMSync{
		driveClient: driveClient,
	}, nil
}

// ShouldSync は同期対象のカテゴリ・サブカテゴリかどうかを判定（除外ベース）
func (ns *NotebookLMSync) ShouldSync(category, subCategory string) bool {
	// カテゴリ全体が除外されているか
	if config.NotebookLMSyncExcludeCategories[category] {
		return false
	}
	// サブカテゴリが除外されているか
	if excludedSubs, ok := config.NotebookLMSyncExcludeSubCategories[category]; ok {
		for _, sub := range excludedSubs {
			if sub == subCategory {
				return false
			}
		}
	}
	// NotebookLMカテゴリマッピングが存在するか
	if _, exists := config.NotebookLMCategoryMap[category]; !exists {
		return false
	}
	return true
}

// SyncFile はファイルをNotebookLMに同期
func (ns *NotebookLMSync) SyncFile(ctx context.Context, fileID, fileName, notebookCategory, ocrText string, facts []string, summary, dateStr string, fiscalYear int) error {
	// 日付をフォーマット
	formattedDate := formatDateForNotebook(dateStr)

	// 順次処理を保証するためロックを取得
	ns.mu.Lock()
	defer ns.mu.Unlock()

	// 累積ドキュメントを取得または作成
	docID, mimeType, err := ns.getOrCreateAccumulatedDoc(ctx, fiscalYear, notebookCategory)
	if err != nil {
		return fmt.Errorf("累積ドキュメント取得/作成失敗: %w", err)
	}

	// ドキュメントに追記
	entryText := ns.formatEntry(formattedDate, fileName, fileID, ocrText, facts, summary, notebookCategory)
	if err := ns.appendToDoc(ctx, docID, mimeType, entryText); err != nil {
		return fmt.Errorf("ドキュメント追記失敗: %w", err)
	}

	// 元ファイルに同期済みマーカーを設定
	ns.markAsSynced(ctx, fileID)

	log.Printf("NotebookLM同期完了: %s → %d年度_%s", fileName, fiscalYear, notebookCategory)
	return nil
}

// formatDateForNotebook はYYYYMMDD形式をYYYY/MM/DD形式に変換
func formatDateForNotebook(dateStr string) string {
	if len(dateStr) != 8 {
		return time.Now().Format("2006-01-02")
	}
	// 要件 YYYY-MM-DD
	return fmt.Sprintf("%s-%s-%s", dateStr[:4], dateStr[4:6], dateStr[6:8])
}

// formatEntry はエントリテキストを要件に基づきフォーマット
func (ns *NotebookLMSync) formatEntry(formattedDate, fileName, fileID, ocrText string, facts []string, summary, notebookCategory string) string {
	fileURL := fmt.Sprintf("https://drive.google.com/file/d/%s/view", fileID)

	// facts を文字列に変換
	factsStr := ""
	if len(facts) > 0 {
		for _, fact := range facts {
			factsStr += fmt.Sprintf("- %s\n", fact)
		}
	} else {
		factsStr = "- （抽出なし）\n"
	}

	// summary部分（空なら省略）
	summarySection := ""
	if summary != "" {
		summarySection = fmt.Sprintf("\n要約（自動・要確認）:\n%s\n", summary)
	}

	return fmt.Sprintf(`---
## %s

カテゴリ: %s
最終更新: %s
元ファイル: %s

重要情報（抽出・推測なし）:
%s%s
本文（OCR原文）:
%s

`, fileName, notebookCategory, formattedDate, fileURL, factsStr, summarySection, ocrText)
}

// getOrCreateAccumulatedDoc は年度別・カテゴリ別統合ドキュメントを取得または作成
func (ns *NotebookLMSync) getOrCreateAccumulatedDoc(ctx context.Context, fiscalYear int, notebookCategory string) (string, string, error) {
	syncFolderID := config.FolderIDs["NOTEBOOKLM_SYNC"]
	if syncFolderID == "" {
		return "", "", fmt.Errorf("NOTEBOOKLM_SYNCフォルダIDが設定されていません")
	}

	docName := fmt.Sprintf("%d年度_%s", fiscalYear, notebookCategory)

	// 既存のドキュメントを検索
	docID, mimeType, err := ns.findDocByName(ctx, docName, syncFolderID)
	if err != nil {
		return "", "", err
	}
	if docID != "" {
		return docID, mimeType, nil
	}

	// 新規作成
	docID, err = ns.createUnifiedDoc(ctx, docName, syncFolderID, fiscalYear, notebookCategory)
	if err != nil {
		return "", "", err
	}

	return docID, "application/vnd.google-apps.document", nil
}

// findDocByName はフォルダ内でファイルを名前で検索し、IDとMimeTypeを返す
func (ns *NotebookLMSync) findDocByName(ctx context.Context, docName, parentID string) (string, string, error) {
	query := fmt.Sprintf("name='%s' and '%s' in parents and trashed=false", docName, parentID)

	fileList, err := ns.driveClient.service.Files.List().
		Q(query).
		Fields("files(id, name, mimeType)").
		SupportsAllDrives(true).
		IncludeItemsFromAllDrives(true).
		Context(ctx).
		Do()
	if err != nil {
		return "", "", fmt.Errorf("ドキュメント検索エラー: %w", err)
	}

	if len(fileList.Files) > 0 {
		return fileList.Files[0].Id, fileList.Files[0].MimeType, nil
	}
	return "", "", nil
}

// createUnifiedDoc は新しい統合ドキュメント（Google Doc）を作成
func (ns *NotebookLMSync) createUnifiedDoc(ctx context.Context, docName, parentID string, fiscalYear int, notebookCategory string) (string, error) {
	// Google Doc として作成
	file := &drive.File{
		Name:     docName,
		MimeType: "application/vnd.google-apps.document",
		Parents:  []string{parentID},
	}

	createdDoc, err := ns.driveClient.oauthDriveService.Files.Create(file).
		Fields("id").
		SupportsAllDrives(true).
		Context(ctx).
		Do()
	if err != nil {
		return "", fmt.Errorf("ドキュメント作成失敗: %w", err)
	}

	docID := createdDoc.Id

	// 初期ヘッダーを挿入
	headerText := fmt.Sprintf("# %d年度 %s\n\nこのファイルは NotebookLM 用に自動生成された書類OCRテキストの統合ファイルです。\n\n", fiscalYear, notebookCategory)

	requests := []*docs.Request{
		{
			InsertText: &docs.InsertTextRequest{
				Location: &docs.Location{Index: 1},
				Text:     headerText,
			},
		},
	}
	_, err = ns.driveClient.docsService.Documents.BatchUpdate(docID, &docs.BatchUpdateDocumentRequest{
		Requests: requests,
	}).Context(ctx).Do()
	if err != nil {
		log.Printf("初期ヘッダー挿入失敗 (ID: %s): %v", docID, err)
		// 致命的ではないとして続行
	}

	// オーナー権限を転送（設定されている場合）
	if config.NotebookLMOwnerEmail != "" {
		permission := &drive.Permission{
			Role:         "owner",
			Type:         "user",
			EmailAddress: config.NotebookLMOwnerEmail,
		}
		_, err := ns.driveClient.oauthDriveService.Permissions.Create(docID, permission).
			TransferOwnership(true).
			Context(ctx).
			Do()
		if err != nil {
			log.Printf("オーナー権限の転送に失敗しました（容量制限に注意）: %v", err)
		} else {
			log.Printf("ファイルのオーナー権限を転送しました: %s", config.NotebookLMOwnerEmail)
		}
	}

	log.Printf("統合ドキュメント作成: %s (ID: %s)", docName, docID)
	return docID, nil
}

// appendToDoc はファイルの末尾にテキストを追記 (Docs APIを使用)
func (ns *NotebookLMSync) appendToDoc(ctx context.Context, docID, mimeType, text string) error {
	if mimeType != "application/vnd.google-apps.document" {
		return fmt.Errorf("非対応のMIMEタイプです: %s", mimeType)
	}

	maxRetries := 5
	for attempt := 0; attempt < maxRetries; attempt++ {
		// EndOfSegmentLocation を使用して末尾に追記（より安定）
		requests := []*docs.Request{
			{
				InsertText: &docs.InsertTextRequest{
					EndOfSegmentLocation: &docs.EndOfSegmentLocation{},
					Text:                 text,
				},
			},
		}

		// 実行
		_, err := ns.driveClient.docsService.Documents.BatchUpdate(docID, &docs.BatchUpdateDocumentRequest{
			Requests: requests,
		}).Context(ctx).Do()

		if err == nil {
			return nil
		}

		if ns.isRetryable(err) && attempt < maxRetries-1 {
			log.Printf("追記リトライ中 (%d/%d): %v", attempt+1, maxRetries, err)
			ns.backoff(attempt)
			continue
		}

		return fmt.Errorf("ドキュメント追記失敗: %w", err)
	}

	return fmt.Errorf("最大リトライ回数を超えました")
}

// isRetryable はリトライすべきエラーかどうかを判定
func (ns *NotebookLMSync) isRetryable(err error) bool {
	if gerr, ok := err.(*googleapi.Error); ok {
		return gerr.Code == 409 || gerr.Code == 429 || gerr.Code >= 500
	}
	return false
}

// backoff は指数バックオフ
func (ns *NotebookLMSync) backoff(attempt int) {
	duration := time.Duration(1<<uint(attempt)) * time.Second
	duration += time.Duration(rand.Intn(1000)) * time.Millisecond
	time.Sleep(duration)
}

// markAsSynced はファイルを同期済みとしてマーク
func (ns *NotebookLMSync) markAsSynced(ctx context.Context, fileID string) {
	file := &drive.File{
		Properties: map[string]string{
			processedMarker: "true",
		},
	}

	_, err := ns.driveClient.service.Files.Update(fileID, file).
		SupportsAllDrives(true).
		Context(ctx).
		Do()
	if err != nil {
		log.Printf("同期マーキングエラー: %v", err)
	}
}

// IsAlreadySynced はファイルが既に同期済みかチェック
func (ns *NotebookLMSync) IsAlreadySynced(ctx context.Context, fileID string) bool {
	file, err := ns.driveClient.service.Files.Get(fileID).
		Fields("properties").
		SupportsAllDrives(true).
		Context(ctx).
		Do()
	if err != nil {
		log.Printf("同期状態チェックエラー: %v", err)
		return false
	}

	return file.Properties[processedMarker] == "true"
}
